File: engine/src/move_ordering.rs
use std::collections::HashMap;

use crate::types::*;

/// Maximum search depth for killer move storage.
const MAX_DEPTH: usize = 64;

/// Number of killer move slots per depth.
const KILLER_SLOTS: usize = 2;

/// Orders moves to improve alpha-beta pruning efficiency.
///
/// Uses three heuristics:
/// 1. **MVV-LVA** (Most Valuable Victim - Least Valuable Attacker):
///    Prioritizes actions that interact with high-value targets.
/// 2. **Killer moves**: Remembers moves that caused beta cutoffs at the
///    same depth in sibling nodes.
/// 3. **History heuristic**: Tracks cumulative success of each move
///    across the entire search.
#[derive(Debug, Clone)]
pub struct MoveOrderer {
    /// Killer move slots: [depth][slot] -> action key.
    killer_moves: Vec<[Option<String>; KILLER_SLOTS]>,
    /// History table: action_key -> cumulative depth^2 bonus.
    history_table: HashMap<String, f64>,
    /// Counter for total ordering calls (diagnostic).
    ordering_calls: u64,
}

impl MoveOrderer {
    pub fn new() -> Self {
        Self {
            killer_moves: vec![[None, None]; MAX_DEPTH],
            history_table: HashMap::new(),
            ordering_calls: 0,
        }
    }

    /// Sort `actions` in-place so the most promising moves come first.
    ///
    /// The scoring combines MVV-LVA, killer bonuses, and history scores,
    /// then sorts descending.
    pub fn order_moves(
        &self,
        actions: &mut [ExecutionAction],
        state: &OnChainState,
        depth: u32,
    ) {
        if actions.len() <= 1 {
            return;
        }

        // Compute a score for each action and sort by it descending.
        let mut scored: Vec<(usize, f64)> = actions
            .iter()
            .enumerate()
            .map(|(i, action)| {
                let mvv = self.mvv_lva_score(action, state);
                let killer = self.killer_bonus(action, depth);
                let history = self.history_score(action);
                (i, mvv + killer + history)
            })
            .collect();

        scored.sort_by(|a, b| b.1.partial_cmp(&a.1).unwrap_or(std::cmp::Ordering::Equal));

        // Reorder actions according to scored order.
        let ordered: Vec<ExecutionAction> = scored
            .iter()
            .map(|(idx, _)| actions[*idx].clone())
            .collect();

        for (i, action) in ordered.into_iter().enumerate() {
            actions[i] = action;
        }
    }

    /// Record a move as a killer at the given depth.
    ///
    /// Uses a two-slot replacement scheme: if the move is not already in
    /// slot 0, shift slot 0 to slot 1 and insert the new move in slot 0.
    pub fn update_killer(&mut self, depth: u32, action: &ExecutionAction) {
        let d = depth as usize;
        if d >= MAX_DEPTH {
            return;
        }

        let key = action.action_key();
        let slots = &mut self.killer_moves[d];

        // Don't add if already in slot 0
        if slots[0].as_ref() == Some(&key) {
            return;
        }

        // Shift slot 0 -> slot 1, insert new in slot 0
        slots[1] = slots[0].take();
        slots[0] = Some(key);
    }

    /// Record a successful move in the history table.
    ///
    /// Bonus = depth^2, so deeper cutoffs are weighted more heavily.
    pub fn update_history(&mut self, action: &ExecutionAction, depth: u32) {
        let key = action.action_key();
        let bonus = (depth as f64) * (depth as f64);
        let entry = self.history_table.entry(key).or_insert(0.0);
        *entry += bonus;

        // Age / cap the history to prevent overflow domination
        if *entry > 100_000.0 {
            // Scale all entries down
            for val in self.history_table.values_mut() {
                *val *= 0.5;
            }
        }
    }

    /// MVV-LVA heuristic: score an action based on the "value" of the
    /// target and the "cost" of the attacker.
    ///
    /// For on-chain actions:
    /// - High-value: Liquidations (capturing collateral), large swaps
    /// - Low-cost: Actions with low priority fees
    ///
    /// Returns a score in roughly [0, 100].
    pub fn mvv_lva_score(&self, action: &ExecutionAction, state: &OnChainState) -> f64 {
        let victim_value = self.estimate_victim_value(action, state);
        let attacker_cost = self.estimate_attacker_cost(action);

File: src/engine/move-ordering.ts
/**
 * MNMX Move Ordering
 *
 * Dramatically improves alpha-beta pruning efficiency by evaluating the
 * most promising actions first.  Combines three heuristics drawn from
 * game-engine research and adapted to on-chain DeFi context:
 *
 *  1. Killer moves   – refutation moves that caused a beta cutoff at the
 *                      same depth in a sibling branch.
 *  2. History table  – actions that have historically produced high scores
 *                      are tried earlier regardless of depth.
 *  3. MVV-LVA        – Most Valuable Victim / Least Valuable Aggressor,
 *                      re-interpreted for DeFi as "highest expected value
 *                      per unit of cost (gas + slippage)".
 */

import type { ExecutionAction, OnChainState } from '../types/index.js';

// ── Killer-Move Storage ─────────────────────────────────────────────

const MAX_KILLER_SLOTS = 2; // two killer slots per depth level

function actionKey(a: ExecutionAction): string {
  return `${a.kind}:${a.pool}:${a.tokenMintIn}:${a.tokenMintOut}:${a.amount}`;
}

// ── Class ───────────────────────────────────────────────────────────

export class MoveOrderer {
  /** killerMoves[depth] holds up to MAX_KILLER_SLOTS action keys. */
  private readonly killerMoves: Map<number, string[]> = new Map();

  /** Cumulative score for every action key that ever caused an improvement. */
  private readonly historyScores: Map<string, number> = new Map();

  // ── Public API ──────────────────────────────────────────────────

  /**
   * Return `actions` sorted so that the most promising candidates
   * appear first.  Does not mutate the input array.
   */
  orderMoves(
    actions: ReadonlyArray<ExecutionAction>,
    state: OnChainState,
    depth: number = 0,
  ): ExecutionAction[] {
    // Score each action with a composite ordering value
    const scored = actions.map((action) => ({
      action,
      orderScore: this.computeOrderScore(action, state, depth),
    }));

    // Sort descending – highest score first
    scored.sort((a, b) => b.orderScore - a.orderScore);

    return scored.map((s) => s.action);
  }

  /**
   * Record a killer move at the given depth.  Called when an action
   * causes a beta cutoff during search.
   */
  updateKillerMove(depth: number, action: ExecutionAction): void {
    const key = actionKey(action);
    let slots = this.killerMoves.get(depth);
    if (!slots) {
      slots = [];
      this.killerMoves.set(depth, slots);
    }

    // Avoid duplicates
    if (slots.includes(key)) return;

    // Shift older killer out if at capacity
    if (slots.length >= MAX_KILLER_SLOTS) {
      slots.shift();
    }
    slots.push(key);
  }

  /**
   * Bump the history score for an action.  Weighted by depth squared
   * so that refutations found deeper in the tree carry more weight.
   */
  updateHistory(action: ExecutionAction, depth: number): void {
    const key = actionKey(action);
    const prev = this.historyScores.get(key) ?? 0;
    this.historyScores.set(key, prev + depth * depth);
  }

  /** Reset all heuristic data (call between unrelated searches). */
  reset(): void {
    this.killerMoves.clear();
    this.historyScores.clear();
  }

  // ── Private ─────────────────────────────────────────────────────

  private computeOrderScore(
    action: ExecutionAction,
    state: OnChainState,
    depth: number,
  ): number {
    let score = 0;

    // 1. Killer-move bonus (highest priority)
    const key = actionKey(action);
    const killers = this.killerMoves.get(depth);
    if (killers?.includes(key)) {
      score += 50_000;
    }

    // 2. History heuristic
    const historyVal = this.historyScores.get(key) ?? 0;
    score += Math.min(historyVal, 30_000); // cap to avoid domination

    // 3. MVV-LVA adapted for DeFi
    score += this.mvvLvaScore(action, state);

    // 4. Explicit priority hint from the action itself
    score += action.priority * 100;

    return score;
  }

  /**
   * MVV-LVA: favour actions that extract the most value (large amounts
   * through liquid pools) with the least cost (low slippage, low gas).
   *
   * "Victim" = value captured (swap output, liquidation bonus, etc.)
   * "Aggressor" = cost to execute (slippage + fees)
   */
  private mvvLvaScore(action: ExecutionAction, state: OnChainState): number {
    // Approximate "victim value" from the action amount
    const amountScore = amountMagnitude(action.amount);

    // Approximate "aggressor cost" from slippage setting
    const costPenalty = action.slippageBps / 10; // higher slippage = more cost

    // Pool liquidity bonus – deeper pools get a boost
    const pool = state.poolStates.get(action.pool);
    let liquidityBonus = 0;
    if (pool) {
      const totalReserves = Number(pool.reserveA + pool.reserveB);
      liquidityBonus = Math.min(Math.log10(totalReserves + 1) * 500, 5_000);
    }

    // Action-kind weighting – some kinds are inherently more valuable
    const kindWeight = ACTION_KIND_WEIGHTS[action.kind] ?? 1;

    return (amountScore * kindWeight + liquidityBonus) - costPenalty;
  }
}

// ── Helpers ─────────────────────────────────────────────────────────

/** Map action amount to a bounded score using log scale. */
function amountMagnitude(amount: bigint): number {
  if (amount <= 0n) return 0;
  // Use string length as a fast proxy for log10
  const digits = amount.toString().length;
  return digits * 1_000;
}

/** Relative importance of each action kind for move ordering. */
const ACTION_KIND_WEIGHTS: Record<string, number> = {
  liquidate: 3,
  swap: 2,
  provide_liquidity: 1.5,
  remove_liquidity: 1.5,
  stake: 1.2,
  unstake: 1.2,
  transfer: 1,
  borrow: 1.3,
  repay: 1.1,
};
